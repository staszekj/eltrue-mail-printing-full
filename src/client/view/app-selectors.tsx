import {TRootState} from "../redux/root-reducer";
import {EEndpointStatus} from "../../common/endpoints";

export const getSearchComponentInputString = (state: TRootState) => state.searchComponent.inputString;
export const getFoundImages = (state: TRootState) => state.searchEndpoint.response;
export const getFoundImagesLenght = (state: TRootState) => state.searchEndpoint.response.length;


export const isOneImageShow = (state: TRootState) => state.oneImageComponent.show;
export const getOneImage = (state: TRootState) => getFoundImagesLenght(state) > 0 ?
    getFoundImages(state)[state.oneImageComponent.arrayId % getFoundImagesLenght(state)]
    : null;

export const getArrayId = (state: TRootState) => state.oneImageComponent.arrayId;
export const getPrevArrayId = (state: TRootState) => (state.oneImageComponent.arrayId - 1) % getFoundImagesLenght(state);
export const getNextArrayId = (state: TRootState) => (state.oneImageComponent.arrayId + 1) % getFoundImagesLenght(state);
export const getNextArrayIdAfterDelete = (state: TRootState) => getFoundImagesLenght(state) > 1 && getFoundImagesLenght(state) - 1 > getArrayId(state) ?
    getArrayId(state) : 0;
export const isShowAfterDelete = (state: TRootState) => getFoundImagesLenght(state) > 1 ? isOneImageShow(state) : false;

export const isAuthorUpdateRequestPending = (state: TRootState) => state.authorUpdateEndpoint.status === EEndpointStatus.PENDING;

export const getTheme = (state: TRootState) => state.theme;
