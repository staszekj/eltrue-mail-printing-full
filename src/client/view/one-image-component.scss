.one-image-component {
  @keyframes wait-spinner {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }

  display: flex;
  justify-content: center;
  align-items: center;

  .toolbar {
    position: absolute;
    right: 0;
    padding-right: 10px;
    display: flex;
    flex-direction: column;

    .close-icon,
    .edit-icon,
    .delete-icon {
      font-size: 32px;
      color: black;
      right: 0;
      top: 0;
      z-index: 3;
      cursor: pointer;
    }
  }

  .spinner-icon {
    position: relative;
    width: 32px;
    height: 48px;

    .spinner::before {
      position: absolute;
      height: 32px;
      width: 32px;
      top: 10px;
      animation: 1.5s linear infinite wait-spinner;
      border: solid 3px #eee;
      border-bottom-color: red;
      border-radius: 50%;
      content: "";
    }
  }

  .one-image {
    position: relative;
    $anim-speed: 0.3s;
    padding: 10px;

    margin: 10px;
    text-align: left;
    opacity: .99;
    overflow: hidden;
    border-radius: 3px;
    box-shadow: 0 0 20px 0 rgba(0, 0, 0, .05);

    img {
      width: 90vw;
      height: auto;
      max-width: none;
      max-height: 90vh;
      object-fit: contain;
    }

    .details {
      font-size: 16px;
      padding: 20px;
      color: black;
      position: absolute;
      bottom: 0;
      left: 0;
      z-index: 3;

      .details-content {
        display: block;
        opacity: 0;
        position: relative;
        top: 25px;
        transition-property: top, opacity;
        transition-duration: $anim-speed;
        transition-delay: 0s;
      }

      .title {
        line-height: 1;
        opacity: 1;
        font-weight: 600;
        font-size: 18px;
      }

      .info {
        line-height: 1.2;
        margin-top: 5px;
        font-size: 12px;
      }
    }

    .right-toolbar {
      font-size: 16px;
      padding: 20px;
      color: black;
      position: absolute;
      top: 50%;
      margin-top: -100px;
      right: 0;
      z-index: 3;

      .right-toolbar-content {
        display: block;
        opacity: 0;
        position: relative;
        left: 50px;
        transition-property: left, opacity;
        transition-duration: $anim-speed;
        transition-delay: 0s;
      }

      .icon-1 {
        color: black;
        font-weight: 600;
        font-size: 70px;
        cursor: pointer;
      }
    }

    .left-toolbar {
      font-size: 16px;
      padding: 20px;
      color: black;
      position: absolute;
      top: 50%;
      margin-top: -100px;
      left: 0;
      z-index: 3;

      .left-toolbar-content {
        display: block;
        opacity: 0;
        position: relative;
        right: 50px;
        transition-property: right, opacity;
        transition-duration: $anim-speed;
        transition-delay: 0s;
      }

      .icon-1 {
        color: black;
        font-weight: 600;
        font-size: 70px;
        cursor: pointer;
      }
    }

    &:focus,
    &:hover {

      &:before,
      .details-content,
      .right-toolbar-content,
      .left-toolbar-content {
        opacity: 1;
      }

      .details-content {
        top: 0;
      }

      .title {
        transition-delay: 0.15s;
      }

      .info {
        transition-delay: 0.25s;
      }

      .right-toolbar-content {
        left: 0;
      }

      .left-toolbar-content {
        right: 0;
      }

      .icon-1 {
        transition-delay: 0.15s;
      }

      .icon-2 {
        transition-delay: 0.25s;
      }
    }
  }
}
