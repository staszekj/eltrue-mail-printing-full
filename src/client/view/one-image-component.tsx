import React, {FunctionComponent, useEffect, useRef, useState} from 'react'
import classnames from 'classnames';
import {MdArrowBack, MdArrowForward, MdClose, MdEdit, MdDelete} from 'react-icons/md';
import {useDispatch, useSelector} from 'react-redux';

import "./one-image-component.scss"
import {TImageMeta} from "../../common/search-endpoint";
import {
    authorUpdateAsyncAction,
    deleteImageAsyncAction,
    oneImageComponentBackwardClickAction,
    oneImageComponentCloseClickAction,
    oneImageComponentForwardClickAction
} from "./app-actions";
import {
    getNextArrayId,
    getNextArrayIdAfterDelete,
    getPrevArrayId,
    isAuthorUpdateRequestPending,
    isShowAfterDelete
} from "./app-selectors";
import {EditableText} from "./widgets/editable-text-widget";

export interface TOneImageComponentProp {
    image: TImageMeta;
}

export const OneImageComponent: FunctionComponent<TOneImageComponentProp> = ({image}) => {

    const imgRef = useRef<HTMLImageElement>(null);

    const [isAuthorEditMode, setAuthorEditMode] = useState<boolean>(false);
    const [isImageFullyLoaded, setImageFullyLoaded] = useState<boolean>(false);
    const [currentAuthor, setCurrentAuthor] = useState<string>(image.author);
    const [currentTimeout, setCurrentTimeout] = useState<ReturnType<typeof setTimeout> | null>(null);
    const [currentId, setCurrentId] = useState<string>(image.id);

    const prevArrayId = useSelector(getPrevArrayId);
    const nextArrayId = useSelector(getNextArrayId);
    const arrayIdAfterDelete = useSelector(getNextArrayIdAfterDelete);
    const showAfterDelete = useSelector(isShowAfterDelete);
    const isRequestPending = useSelector(isAuthorUpdateRequestPending);

    const dispatch = useDispatch();

    const onCloseBtnClick = () => dispatch(oneImageComponentCloseClickAction());
    const onDeleteBtnClickHandler = () => dispatch(deleteImageAsyncAction.request({
        id: image.id,
        arrayId: arrayIdAfterDelete,
        show: showAfterDelete
    }));
    const onForwardBtnClick = () => dispatch(oneImageComponentForwardClickAction({arrayId: nextArrayId}));
    const onPrevBtnClick = () => dispatch(oneImageComponentBackwardClickAction({arrayId: prevArrayId}));

    const openEditMode = () => {
        if (isAuthorEditMode){
            return;
        }
        setCurrentAuthor(image.author);
        setAuthorEditMode(true);
    };
    const onEnterTitle = () => {
        setAuthorEditMode(false);
        dispatch(authorUpdateAsyncAction.request({id: image.id, author: currentAuthor}));
    };

    const editableTextString = isAuthorEditMode || isRequestPending ? currentAuthor : image.author;
    const editButtonWithClickHandlerEl = (<div key={"editButtonWithClickHandler"} className={classnames("edit-icon")} onClick={() => openEditMode()}><MdEdit/></div>);
    const editButtonWithoutClickHandlerEl = (<div key={"editButtonWithoutClickHandler"} className={classnames("edit-icon")}><MdEdit/></div>);
    const editButtonEl = isAuthorEditMode ? editButtonWithoutClickHandlerEl : editButtonWithClickHandlerEl;
    const spinnerEl = !isImageFullyLoaded ? (
        <div className={classnames("spinner-icon")}>
            <div className={"spinner"}/>
        </div>) : null;

    if (currentId !== image.id) {
        setAuthorEditMode(false);
        setImageFullyLoaded(false);
        setCurrentId(image.id);
    }

    useEffect(() => {
        if (currentTimeout) {
            clearTimeout(currentTimeout)
        }
        const delayedFullImageLoading = setTimeout(() => {
            const imageLoader = new Image();
            imageLoader.src = image.downloadUrl;
            imageLoader.onload = () => {
                const imageRef = imgRef.current;
                if (imageRef && imageRef.src === image.imageV300Url) {
                    imageRef.src = imageLoader.src;
                    setImageFullyLoaded(true);
                }
            };
        }, 500);
        setCurrentTimeout(delayedFullImageLoading)
    }, [image.id]); // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <div className={"one-image-component"}>
            <div className={"one-image"}>
                <div className={classnames("toolbar")}>
                    <div className={classnames("close-icon")} onClick={() => onCloseBtnClick()}><MdClose/></div>
                    {editButtonEl}
                    <div className={classnames("delete-icon")} onClick={() => onDeleteBtnClickHandler()}><MdDelete/>
                    </div>
                    {spinnerEl}
                </div>
                <img ref={imgRef} src={image.imageV300Url} alt={image.author}/>
                <div className={"details"}>
                    <EditableText text={editableTextString} isInput={isAuthorEditMode || isRequestPending}
                                  isReadOnly={isRequestPending}
                                  isSpinner={isRequestPending}
                                  classNames={classnames("details-content", "title")}
                                  onEnter={onEnterTitle}
                                  onChange={(text) => setCurrentAuthor(text)}
                                  onTextClick={() => openEditMode()}>
                    </EditableText>
                    <span className={classnames("details-content", "info")}>{`${image.width} x ${image.height}`}</span>
                </div>
                <div className={"right-toolbar"}>
                    <div className={classnames("right-toolbar-content", "icon-1")} onClick={() => onForwardBtnClick()}>
                        <MdArrowForward/></div>
                </div>
                <div className={"left-toolbar"}>
                    <div className={classnames("left-toolbar-content", "icon-1")} onClick={() => onPrevBtnClick()}>
                        <MdArrowBack/></div>
                </div>
            </div>
        </div>
    );
};
