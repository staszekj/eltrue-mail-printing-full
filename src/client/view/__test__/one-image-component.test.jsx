import * as React from 'react';
import {shallow} from 'enzyme'
import {useDispatch, useSelector} from 'react-redux';
import {getNextArrayId, getPrevArrayId, isAuthorUpdateRequestPending, getNextArrayIdAfterDelete, isShowAfterDelete} from "../app-selectors";
import {OneImageComponent} from "../one-image-component";

jest.mock('react-redux', () => {
    return {
        useSelector: jest.fn(),
        useDispatch: jest.fn()
    }
});

describe('<OneImageComponent/>', () => {

    const useRefSpy = jest.spyOn(React, "useRef");
    const useStateSpy = jest.spyOn(React, "useState");
    const useEffectSpy = jest.spyOn(React, "useEffect");

    const setStateSpy = jest.fn();
    const dispatchSpy = jest.fn();

    const useSelectorMap = {
        [getNextArrayId]: 1,
        [getPrevArrayId]: 2,
        [isAuthorUpdateRequestPending]: false,
        [getNextArrayIdAfterDelete]: 0,
        [isShowAfterDelete]: false
    };

    beforeEach(() => {
        jest.useFakeTimers();

        useRefSpy.mockReturnValue({current: {}} );
        useStateSpy.mockImplementation((initValue) => [initValue, setStateSpy]);
        useSelector.mockImplementation((selectorfn) => useSelectorMap[selectorfn]);
        useDispatch.mockImplementation(() => dispatchSpy);

        jest.clearAllMocks();
    });

    const props = {
        image: {
            id: "abc",
            author: "fake_author",
            width: 1280,
            height: 1024
        }
    };

    it('should be rendered', () => {
        //when
        const oneImageComponentEl = shallow(<OneImageComponent {...props}/>);

        //then
        expect(oneImageComponentEl).toMatchSnapshot();
    });

    it('should handle useEffect', () => {
        //TODO
        //To mock "new Image()" we should create "export const getNewImage = () => new Image()"
        //function in other module
        //then that module should be mocked here

        //given
        shallow(<OneImageComponent {...props}/>);
        const useEffectHandleFn = useEffectSpy.mock.calls[0][0];

        //when
        useEffectHandleFn();
        const setTimeoutHandler = setTimeout.mock.calls[0][0];
        setTimeoutHandler();

        //then
        expect(setStateSpy).toHaveBeenCalledWith(expect.any(Number));
    })
});
