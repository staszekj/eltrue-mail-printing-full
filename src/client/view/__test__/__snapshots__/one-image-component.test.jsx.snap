// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OneImageComponent/> should be rendered 1`] = `
<div
  className="one-image-component"
>
  <div
    className="one-image"
  >
    <div
      className="toolbar"
    >
      <div
        className="close-icon"
        onClick={[Function]}
      >
        <MdClose />
      </div>
      <div
        className="edit-icon"
        onClick={[Function]}
      >
        <MdEdit />
      </div>
      <div
        className="delete-icon"
        onClick={[Function]}
      >
        <MdDelete />
      </div>
      <div
        className="spinner-icon"
      >
        <div
          className="spinner"
        />
      </div>
    </div>
    <img
      alt="fake_author"
    />
    <div
      className="details"
    >
      <EditableText
        classNames="details-content title"
        isInput={false}
        isReadOnly={false}
        isSpinner={false}
        onChange={[Function]}
        onEnter={[Function]}
        onTextClick={[Function]}
        text="fake_author"
      />
      <span
        className="details-content info"
      >
        1280 x 1024
      </span>
    </div>
    <div
      className="right-toolbar"
    >
      <div
        className="right-toolbar-content icon-1"
        onClick={[Function]}
      >
        <MdArrowForward />
      </div>
    </div>
    <div
      className="left-toolbar"
    >
      <div
        className="left-toolbar-content icon-1"
        onClick={[Function]}
      >
        <MdArrowBack />
      </div>
    </div>
  </div>
</div>
`;
